# Chapter 3: Data exploration

In [1]:
%reset
low_memory=False
import numpy as np
import pandas as pd

Once deleted, variables cannot be recovered. Proceed (y/[n])?  y


## 3.1 Introduction & Problem Setting

Today you will be tackeling your very first data cleaning task! You will continue your work on the titanic dataset and exploring the many approaches you can have when cleaning a dataset. You will discover how each approach has its own advantages and disadvantages.

## 3.2 Explore the content of the dataset

Load in the titanic dataset and start from the unmodified data. Explore the dataset and identify the following things:

- How many records and columns do you have?
- How many columns contain missing values?
- How many columns do you want to transform?
- Which columns are these?

In [2]:
titanic = pd.read_csv("titanic_uncleaned.csv", sep = ";")
titanic.shape

(892, 12)

In [3]:
titanic.describe()

PassengerId      Pclass         Age       SibSp       Parch
count   892.000000  891.000000  714.000000  891.000000  891.000000
mean    446.500000    2.308642   29.706120    0.523008    0.381594
std     257.642517    0.836071   14.523986    1.102743    0.806057
min       1.000000    1.000000    0.420000    0.000000    0.000000
25%     223.750000    2.000000   20.125000    0.000000    0.000000
50%     446.500000    3.000000   28.000000    0.000000    0.000000
75%     669.250000    3.000000   38.000000    1.000000    0.000000
max     892.000000    3.000000   80.000000    8.000000    6.000000

In [4]:
titanic.isna().describe()

PassengerId Survived Pclass   Name    Sex    Age  SibSp  Parch Ticket  \
count          892      892    892    892    892    892    892    892    892   
unique           1        2      2      2      1      2      2      2      2   
top          False    False  False  False  False  False  False  False  False   
freq           892      891    891    891    892    714    891    891    891   

         Fare Cabin Embarked  
count     892   892      892  
unique      2     2        2  
top     False  True    False  
freq      891   688      890

In [5]:
titanic.dtypes

PassengerId      int64
Survived           str
Pclass         float64
Name               str
Sex                str
Age            float64
SibSp          float64
Parch          float64
Ticket             str
Fare               str
Cabin              str
Embarked           str
dtype: object

The dataset has 892 records and 12 columns.

Of those 12 columns, only two have no missing values. Those being 'PassengerId' and 'Sex'.

All other columns have only a few missing values, with only 'Age' and 'Cabin' having significantly more missing values.

About half of the columns are already in a number format. The ones we likely want to fix are:

- Survived
- Name
- Sex
- Age
- Ticket
- Cabin
- Embarked

## 3.3 Data transformations 

Transform the data types of your columns! Make sure your data is in the desired from so you can easily handle the missing values later.

*Hint: we have not yet seen what to do with data such as in the columns 'Name', 'Ticket' and 'Cabin'. Thus, you can leave those three columns as they are.*

In [6]:
titanic['Survived'] = titanic['Survived'].map({'No': 0, 'Yes' : 1})
titanic['Sex'] = titanic['Sex'].map({'male': 0, 'female' : 1})
titanic = pd.get_dummies(titanic, columns = ['Embarked'])
titanic['Embarked_C'] = titanic['Embarked_C'].astype('float')
titanic['Embarked_Q'] = titanic['Embarked_Q'].astype('float')
titanic['Embarked_S'] = titanic['Embarked_S'].astype('float')

In [7]:
titanic.head()

PassengerId  Survived  Pclass  \
0            1       0.0     3.0   
1            2       1.0     1.0   
2            3       1.0     3.0   
3            4       1.0     1.0   
4            5       0.0     3.0   

                                                Name  Sex   Age  SibSp  Parch  \
0                            Braund, Mr. Owen Harris    0   NaN    1.0    0.0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...    1  38.0    1.0    0.0   
2                             Heikkinen, Miss. Laina    1  26.0    0.0    0.0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)    1  35.0    1.0    0.0   
4                           Allen, Mr. William Henry    0  35.0    0.0    0.0   

             Ticket     Fare Cabin  Embarked_C  Embarked_Q  Embarked_S  
0         A/5 21171     7.25   NaN         0.0         0.0         1.0  
1          PC 17599  712.833   C85         1.0         0.0         0.0  
2  STON/O2. 3101282    7.925   NaN         0.0         0.0         1.0  
3            113803     53.1  C123         0.0         0.0         1.0  
4            373450     8.05   NaN         0.0         0.0         1.0

## 3.4 Missing values

Now it's time to tackle the missing values. Fist of all, have a look at what you may have already done on this subject. Compare the data you currently have with the raw dataset. Did you already solve some missing values?

In [8]:
titanic.isna().describe()

PassengerId Survived Pclass   Name    Sex    Age  SibSp  Parch Ticket  \
count          892      892    892    892    892    892    892    892    892   
unique           1        2      2      2      1      2      2      2      2   
top          False    False  False  False  False  False  False  False  False   
freq           892      891    891    891    892    714    891    891    891   

         Fare Cabin Embarked_C Embarked_Q Embarked_S  
count     892   892        892        892        892  
unique      2     2          1          1          1  
top     False  True      False      False      False  
freq      891   688        892        892        892

In [9]:
titanic[np.logical_and(np.logical_and(titanic.Embarked_C == 0, titanic.Embarked_Q == 0), titanic.Embarked_S == 0)]

PassengerId  Survived  Pclass                                       Name  \
61            62       1.0     1.0                        Icard, Miss. Amelie   
829          830       1.0     1.0  Stone, Mrs. George Nelson (Martha Evelyn)   

     Sex   Age  SibSp  Parch  Ticket Fare Cabin  Embarked_C  Embarked_Q  \
61     1  38.0    0.0    0.0  113572   80   B28         0.0         0.0   
829    1  62.0    0.0    0.0  113572   80   B28         0.0         0.0   

     Embarked_S  
61          0.0  
829         0.0

By using the .get_dummies() function we have already eliminated some missing values. All three new columns no longer contain missing value, but instead only a 0 or 1. Records where this value was missing have been replaced with a 0 in each column.

In [11]:
titanic['Fare'].dtype

<StringDtype(storage='python', na_value=nan)>

In [18]:
def clean_fare(value):
    if type(value) == 'str':
        if len(value) > 3:
            if value[-3] == '.' and value[-2:].isdigit():
                value = value.replace('.', ',', 1)
            value = value.replace('.', '')
            value = value.replace(',', '.')
        return value
    return value

titanic['Fare'] = titanic['Fare'].apply(clean_fare)

Now it's time to get to work! Take a copy of your cleaned dataset and handle the missing values according the the three ways you've seen. Analyse and describe the effect each method has on the data. How many rows do you retain, how many columns? How good is the data quality? Can you think of ways to improve your approach?

### 3.4.1: Data removal

In [19]:
titanic2 = titanic.dropna()

In [20]:
titanic2.isna().describe()

PassengerId Survived Pclass   Name    Sex    Age  SibSp  Parch Ticket  \
count          185      185    185    185    185    185    185    185    185   
unique           1        1      1      1      1      1      1      1      1   
top          False    False  False  False  False  False  False  False  False   
freq           185      185    185    185    185    185    185    185    185   

         Fare  Cabin Embarked_C Embarked_Q Embarked_S  
count     185    185        185        185        185  
unique      1      1          1          1          1  
top     False  False      False      False      False  
freq      185    185        185        185        185

Simply dropping all the records with missing values leads to a huge loss in data.

Since we know that the 'Cabin' column consists mostly of missing values, we can drop it to save other records.

In [12]:
titanic2 = titanic.drop(columns=['Cabin'])
titanic2 = titanic2.dropna()

In [13]:
titanic2.isna().describe()

PassengerId Survived Pclass   Name    Sex    Age  SibSp  Parch Ticket  \
count          713      713    713    713    713    713    713    713    713   
unique           1        1      1      1      1      1      1      1      1   
top          False    False  False  False  False  False  False  False  False   
freq           713      713    713    713    713    713    713    713    713   

         Fare Embarked_C Embarked_Q Embarked_S  
count     713        713        713        713  
unique      1          1          1          1  
top     False      False      False      False  
freq      713        713        713        713

Now we are retaining 713 records, which is a lot more!

### 3.4.1: Data Imputation

In [14]:
titanic3 = titanic

In [15]:
titanic3.isna().describe()

PassengerId Survived Pclass   Name    Sex    Age  SibSp  Parch Ticket  \
count          892      892    892    892    892    892    892    892    892   
unique           1        2      2      2      1      2      2      2      2   
top          False    False  False  False  False  False  False  False  False   
freq           892      891    891    891    892    714    891    891    891   

         Fare Cabin Embarked_C Embarked_Q Embarked_S  
count     892   892        892        892        892  
unique      2     2          1          1          1  
top     False  True      False      False      False  
freq      891   688        892        892        892

In [16]:
titanic3['Survived'] = titanic3['Survived'].ffill()
titanic3['Pclass'] = titanic3['Pclass'].ffill()
titanic3['Name'] = titanic3['Name'].ffill()
titanic3['Sex'] = titanic3['Sex'].ffill()
titanic3['Age'] = titanic3['Age'].ffill()
titanic3['SibSp'] = titanic3['SibSp'].ffill()
titanic3['Parch'] = titanic3['Parch'].ffill()
titanic3['Ticket'] = titanic3['Ticket'].ffill()
titanic3['Fare'] = titanic3['Fare'].ffill()
titanic3['Cabin'] = titanic3['Cabin'].ffill()

In [17]:
titanic3.isna().describe()

PassengerId Survived Pclass   Name    Sex    Age  SibSp  Parch Ticket  \
count          892      892    892    892    892    892    892    892    892   
unique           1        1      1      1      1      2      1      1      1   
top          False    False  False  False  False  False  False  False  False   
freq           892      892    892    892    892    891    892    892    892   

         Fare  Cabin Embarked_C Embarked_Q Embarked_S  
count     892    892        892        892        892  
unique      1      2          1          1          1  
top     False  False      False      False      False  
freq      892    891        892        892        892

We can see  the 'Age' and 'Cabin' column both still have a missing value for some reason. Let's investigate!

In [18]:
titanic3.loc[np.logical_or(titanic3['Cabin'].isna(), titanic3['Age'].isna())]

PassengerId  Survived  Pclass                     Name  Sex  Age  SibSp  \
0            1       0.0     3.0  Braund, Mr. Owen Harris    0  NaN    1.0   

   Parch     Ticket  Fare Cabin  Embarked_C  Embarked_Q  Embarked_S  
0    0.0  A/5 21171  7.25   NaN         0.0         0.0         1.0

Aha, this makes sense! The first record has missing values, and thus there is no previous one to set the value. We'll loop over the data again with a bfill() to solve this issue.

In [19]:
titanic3['Age'] = titanic3['Age'].bfill()
titanic3['Cabin'] = titanic3['Cabin'].bfill()
titanic3.isna().describe()

PassengerId Survived Pclass   Name    Sex    Age  SibSp  Parch Ticket  \
count          892      892    892    892    892    892    892    892    892   
unique           1        1      1      1      1      1      1      1      1   
top          False    False  False  False  False  False  False  False  False   
freq           892      892    892    892    892    892    892    892    892   

         Fare  Cabin Embarked_C Embarked_Q Embarked_S  
count     892    892        892        892        892  
unique      1      1          1          1          1  
top     False  False      False      False      False  
freq      892    892        892        892        892

### 3.4.3: Data Flagging

Unfortunately python makes copies by reference, which means that even though we have made a copy for our most recent data transformations, we still modified the original dataset. We thus have to load it in again.

In [21]:
titanic = pd.read_csv("titanic_uncleaned.csv", sep = ";")

titanic['Survived'] = titanic['Survived'].map({'No': 0, 'Yes' : 1})
titanic['Sex'] = titanic['Sex'].map({'male': 0, 'female' : 1})
titanic = pd.get_dummies(titanic, columns = ['Embarked'])
titanic['Embarked_C'] = titanic['Embarked_C'].astype('float')
titanic['Embarked_Q'] = titanic['Embarked_Q'].astype('float')
titanic['Embarked_S'] = titanic['Embarked_S'].astype('float')

def clean_fare(value):
    value = str(value)
    if len(value) > 3:
        if value[-3] == '.' and value[-2:].isdigit():
            value = value.replace('.', ',', 1)
        value = value.replace('.', '')
        value = value.replace(',', '.')
    return value

titanic['Fare'] = titanic['Fare'].apply(clean_fare).astype('float')

In [22]:
titanic['Survived'] = titanic['Survived'].fillna(-1)
titanic['Pclass'] = titanic['Pclass'].fillna(-1)
titanic['Name'] = titanic['Name'].fillna("")
titanic['Age'] = titanic['Age'].fillna(-1)
titanic['SibSp'] = titanic['SibSp'].fillna(-1)
titanic['Parch'] = titanic['Parch'].fillna(-1)
titanic['Ticket'] = titanic['Ticket'].fillna("")
titanic['Fare'] = titanic['Fare'].fillna(-1)
titanic['Cabin'] = titanic['Cabin'].fillna("")

In [23]:
titanic.isna().describe()

PassengerId Survived Pclass   Name    Sex    Age  SibSp  Parch Ticket  \
count          892      892    892    892    892    892    892    892    892   
unique           1        1      1      1      1      1      1      1      1   
top          False    False  False  False  False  False  False  False  False   
freq           892      892    892    892    892    892    892    892    892   

         Fare  Cabin Embarked_C Embarked_Q Embarked_S  
count     892    892        892        892        892  
unique      1      1          1          1          1  
top     False  False      False      False      False  
freq      892    892        892        892        892

### 3.5: Custom approach

Combine the knowledge you have gained by comparing the three methods to create one final cleaned dataset where alle missing values are handled in the most optimal way!

In [24]:
titanic = pd.read_csv("titanic_uncleaned.csv", sep = ";")

titanic['Survived'] = titanic['Survived'].map({'No': 0, 'Yes' : 1})
titanic['Sex'] = titanic['Sex'].map({'male': 0, 'female' : 1})
titanic = pd.get_dummies(titanic, columns = ['Embarked'])
titanic['Embarked_C'] = titanic['Embarked_C'].astype('float')
titanic['Embarked_Q'] = titanic['Embarked_Q'].astype('float')
titanic['Embarked_S'] = titanic['Embarked_S'].astype('float')

def clean_fare(value):
    value = str(value)
    if len(value) > 3:
        if value[-3] == '.' and value[-2:].isdigit():
            value = value.replace('.', ',', 1)
        value = value.replace('.', '')
        value = value.replace(',', '.')
    return value

titanic['Fare'] = titanic['Fare'].apply(clean_fare).astype('float')

In [25]:
del titanic['Cabin']

In [26]:
titanic['Survived'] = titanic['Survived'].ffill()
titanic['Pclass'] = titanic['Pclass'].ffill()
titanic['Name'] = titanic['Name'].fillna("")
titanic['Sex'] = titanic['Sex'].ffill()
titanic['Age'] = titanic['Age'].fillna(titanic[titanic['Age'].notna()].Age.median())
titanic['SibSp'] = titanic['SibSp'].fillna(titanic[titanic['SibSp'].notna()].Age.median())
titanic['Parch'] = titanic['Parch'].fillna(titanic[titanic['Parch'].notna()].Age.median())
titanic['Ticket'] = titanic['Ticket'].fillna("")
titanic['Fare'] = titanic['Fare'].fillna(titanic[titanic['Fare'].notna()].Age.median())

In [27]:
titanic.isna().describe()

PassengerId Survived Pclass   Name    Sex    Age  SibSp  Parch Ticket  \
count          892      892    892    892    892    892    892    892    892   
unique           1        1      1      1      1      1      1      1      1   
top          False    False  False  False  False  False  False  False  False   
freq           892      892    892    892    892    892    892    892    892   

         Fare Embarked_C Embarked_Q Embarked_S  
count     892        892        892        892  
unique      1          1          1          1  
top     False      False      False      False  
freq      892        892        892        892

In [28]:
titanic.sample(10)

PassengerId  Survived  Pclass                                      Name  \
714          715       0.0     2.0                     Greenberg, Mr. Samuel   
735          736       0.0     3.0                      Williams, Mr. Leslie   
43            44       1.0     2.0  Laroche, Miss. Simonne Marie Anne Andree   
408          409       0.0     3.0         Birkeland, Mr. Hans Martin Monsen   
250          251       0.0     3.0                    Reed, Mr. James George   
260          261       0.0     3.0                         Smith, Mr. Thomas   
736          737       0.0     3.0   Ford, Mrs. Edward (Margaret Ann Watson)   
289          290       1.0     3.0                      Connolly, Miss. Kate   
178          179       0.0     2.0                        Hale, Mr. Reginald   
834          835       0.0     3.0                    Allum, Mr. Owen George   

     Sex   Age  SibSp  Parch         Ticket       Fare  Embarked_C  \
714    0  52.0    0.0    0.0         250647      13.00         0.0   
735    0  28.5    0.0    0.0          54636     161.00         0.0   
43     1   3.0    1.0    2.0  SC/Paris 2123  415792.00         1.0   
408    0  21.0    0.0    0.0         312992    7775.00         0.0   
250    0  28.0    0.0    0.0         362316       7.25         0.0   
260    0  28.0    0.0    0.0         384461       7.75         0.0   
736    1  48.0    1.0    3.0     W./C. 6608   34375.00         0.0   
289    1  22.0    0.0    0.0         370373       7.75         0.0   
178    0  30.0    0.0    0.0         250653      13.00         0.0   
834    0  18.0    0.0    0.0           2223       8.30         0.0   

     Embarked_Q  Embarked_S  
714         0.0         1.0  
735         0.0         1.0  
43          0.0         0.0  
408         0.0         1.0  
250         0.0         1.0  
260         1.0         0.0  
736         0.0         1.0  
289         1.0         0.0  
178         0.0         1.0  
834         0.0         1.0

In [29]:
titanic.to_csv('titanic_cleaned.csv')